README.md
## Matrices - Matrix Operations and Linear Algebra Library  

[English](https://github.com/moluopro/matrices/blob/master/README.md) &nbsp;&nbsp;&nbsp;[中文](https://github.com/moluopro/matrices/blob/master/README.ZH.md)  

****

&nbsp;&nbsp;&nbsp;&nbsp;Matrices is a matrix operations and linear algebra library for Dart. It was developed to provide matrix operations support for projects such as Astable. All operations are currently implemented in Dart language. To use it, you need to import the matrices.dart file. Also, before using Matrices, you need to understand the concepts related to linear algebra.  

&nbsp;&nbsp;&nbsp;&nbsp;Linear algebra is a branch of mathematics that deals with vectors, linear spaces, linear transformations and finite dimensional systems of linear equations. Linear algebra deals with linear relations. A linear relationship means that the relationship between mathematical objects is expressed in a one-time form. Since nonlinear models in scientific research can often be approximated as linear models, linear algebra is widely used in the natural and social sciences.    

****

## Getting Started  

```dart
  /// Creating a matrix

  var mat = Matrix.fromList([
    [1, 2, 3],
    [4, 5, 6]
  ]);
  print( mat );

  // Matrix: 2x4
  // [1.0, 2.0, 3.0]
  // [4.0, 5.0, 6.0]

  /// Creating a matrix filled with 0’s

  var mat = Matrix.zero(2, 3);
  print( mat );

  // Matrix: 2x3
  // [0.0, 0.0, 0.0]
  // [0.0, 0.0, 0.0]

  /// Creating a matrix filled with 1’s

  var mat = Matrix.one(2, 3);
  print( mat );
  
  // Matrix: 2x3
  // [1.0, 1.0, 1.0]
  // [1.0, 1.0, 1.0]

  /// Creating a matrix filled with certain value

  var mat = Matrix.number(9, 2, 3);
  print( mat );
  
  // Matrix: 2x3
  // [9.0, 9.0, 9.0]
  // [9.0, 9.0, 9.0]

  /// Creating a matrix from a flattened list
  
  // If the length of the flattened list is smaller than the expected matrix size,
  // zeros will be filled in the vacant position.
  
  var mat1 = Matrix.fromFlattenedList([1, 2, 3, 4], 2, 2);
  var mat2 = Matrix.fromFlattenedList([1, 2, 3, 4], 2, 3);
  print(mat1);
  print(mat2);

  // Matrix: 2x2
  // [1.0, 2.0]
  // [3.0, 4.0]

  // Matrix: 2x3
  // [1.0, 2.0, 3.0]
  // [4.0, 0.0, 0.0]

  /// Creating a matrix filled with random values

  var mat = Matrix.random(2, 3);
  print( mat );
  
  // Matrix: 2x3
  // [0.38224693703597046, 0.5412146597032305, 0.6424342222644003]
  // [0.8491145735932242, 0.6236773300386973, 0.25269555696856316]

  /// Addition

  print( mat + 3 );
  print( mat + mat );

  // Matrix: 2x3
  // [4.0, 5.0, 6.0]
  // [7.0, 8.0, 9.0]

  // Matrix: 2x3
  // [2.0, 4.0, 6.0]
  // [8.0, 10.0, 12.0]

  /// Take the matrix mat as example
  Matrix mat 2x4
      [1, 2, 3, 4]
      [5, 6, 7, 8]

  /// Indexing a matrix element
  print( mat[1][2] );

  // 7.0

  /// Number of rows
  print( mat.rowCount );

  // 2

  /// Number of columns
  print( mat.columnCount );

  // 4

  /// Total number of elements
  print( mat.count );

  // 8


  /// Slicing a row

  print( mat.row(0) );

  // An easier slicing method
  print( mat[0] );

  // [1.0, 2.0, 3.0, 4.0]
  // [1.0, 2.0, 3.0, 4.0]

  /// Slicing a column
  print( mat.column(0) );

  // Here, we also provide an easier slicing method for columns
  // Add a minus before real column index to slice column (except the 1st column)

  // Slicing the 1st column
  print( mat[''] );

  // Slicing the 2nd column
  print( mat[-1] );

  // [1.0, 5.0]
  // [1.0, 5.0]
  // [2.0, 6.0]

  /// Transposing
  print( mat.transpose );

  // Matrix: 4x2
  // [1.0, 5.0]
  // [2.0, 6.0]
  // [3.0, 7.0]
  // [4.0, 8.0]

  /// Row Echelon Form
  print( mat.rowEchelonForm );

  // Matrix: 2x4
  // [1.0, 0.0, -1.00000, -2.00000]
  // [0.0, 1.0, 2.00000, 3.00000]


  /// Rank
  print( mat.rank );

  // 2

  /// Take the matrix mat as example, too.
  Matrix mat 2x4
      [1, 2, 3, 4]
      [5, 6, 7, 8]

  /// Modify an element
  mat[0][0] = 6;
  print( mat[0] );

  // [6.0, 2.0, 3.0, 4.0]

  /// Modify a row
  mat.setRow([6, 6, 6, 6], 0);
  print( mat[0] );

  // [6.0, 6.0, 6.0, 6.0]

  // A eaiser way to modify a row
  mat[0] = [8, 8, 8, 8];
  print( mat[0] );

  // [8.0, 8.0, 8.0, 8.0]

  /// Modify a column
  mat.setColumn([6, 6], 0);
  print( mat );

  // Matrix: 2x4
  // [6.0, 2.0, 3.0, 4.0]
  // [6.0, 6.0, 7.0, 8.0]

  // A eaiser way to modify a column
  // Add a minus before real column index to modify column (except the 1st column which requires an empty string)
  mat[''] = [8, 8];
  mat[-1] = [9, 9];
  print( mat );

  // Matrix: 2x4
  // [8.0, 9.0, 3.0, 4.0]
  // [8.0, 9.0, 7.0, 8.0]

  /// Subtraction

  var mat1 = Matrix.fromList([
    [1, 2, 1],
    [1, 2, 1]
  ]);
  print( mat - 3 );
  print( mat - mat1 );

  // Matrix: 2x3
  // [-2.0, -1.0, 0.0]
  // [1.0, 2.0, 3.0]

  // Matrix: 2x3
  // [0.0, 0.0, 2.0]
  // [3.0, 3.0, 5.0]


  /// Scalar Multiplication

  // Note: Scalar should be on the right position of the operator *
  print( mat * 3 );

  // Matrix: 2x3
  // [3.0, 6.0, 9.0]
  // [12.0, 15.0, 18.0]


  /// Matrix Product

  var mat2 = Matrix.fromList([
    [2, 3.5],
    [1, -2],
    [-4, 0.5]
  ]);
  print( mat * mat2 );

  // Matrix: 2x2
  // [-8.0, 1.0]
  // [-11.0, 7.0]

  /// This is only a small part of Matrices, please check the documentation for more usage !

```


## User Guide  
****
Find more functions and usage, please click [Matrices Document](https://abandoft.gitee.io/matrices/) to explore the exciting world of matrices
****

## Who's using it?
- [Astable](https://apps.microsoft.com/store/detail/9NQM9B17MGJS) - A beautiful and powerful structural mechanics software


CHANGELOG.md
## 1.2.7

- doc update

## 1.2.6

- bugs fix

## 1.2.4

- bugs fix

## 1.2.3

- bugs fix

## 1.2.2

- bugs fix
- add doc webpage

## 1.2.1

- bugs fix
- add doc webpage

## 1.2.0

- bugs fix
- big update, please check the documentation

## 1.1.5

- bugs fix
- add rowCount&columnCount&count
- add max&min

## 1.1.4

- bugs fix
- add fromFlattenedList
- add identity matrix

## 1.1.3

- bugs fix

## 1.1.2

- bugs fix
- improve inverse
- methods standardization

## 1.1.0

- bugs fix
- remove emoji methods
- methods standardization

## 1.0.11

- bugs fix
- improve speed of determinant greatly 

## 1.0.9

- bugs fix

## 1.0.8

- update

## 1.0.7

- doc update

## 1.0.6

- doc update

## 1.0.5

- add English Doc

## 1.0.4

- doc update

## 1.0.3

- bugs fix

## 1.0.2

- open source

example/main.dart
import '../lib/matrices.dart';

void main(List<String> args) {
  // Create matrix
  createMatrix();
  createZeros();
  createOnes();
  createNumbers();
  createRandom();

  // Get element, row(s) or column(s)
  queryElement();
  queryRow();
  queryColumn();

  // Transpose, row echelon form & rank
  transpose;
  rowEchelonForm;
  rank;

  // Update element, row(s) or column(s)
  updateElement();
  updateRow();
  updateColumn();

  // Insert row(s) or column(s)
  insertRow();
  insertColumn();
  appendRows();
  appendRows();

  // Delete row(s) or column(s)
  deleteRow();
  deleteColumn();
  deleteRows();
  deleteColumns();

  // Arithmetic operations
  addition();
  subtraction();
  scalarMultiplication();
  matrixMultiplication();

  // Create square matrix
  createSquareMatrix();
  createSquareZeros();
  createSquareOnes();
  createSquareNumbers();
  createSquareRandoms();
  diagFromList();
  diagFromNumber();

  // Get determinant, inverse, eigen value & eigen vector
  det();
  inv();
  eig();
  eigVector();
}

/// Create a matrix instance
void createMatrix() {
  var mat = Matrix.fromList([
    [2, 3, 3, 3],
    [9, 9, 8, 6],
    [1, 1, 2, 9]
  ]);

  print(mat);

  // Matrix: 3x4
  // [2.0, 3.0, 3.0, 3.0]
  // [9.0, 9.0, 8.0, 6.0]
  // [1.0, 1.0, 2.0, 9.0]
}

/// Matrix of zeros
void createZeros() {
  var mat = Matrix.zero(2, 3);

  print(mat);

  // Matrix: 2x3
  // [0.0, 0.0, 0.0]
  // [0.0, 0.0, 0.0]
}

/// Matrix of zeros
void createOnes() {
  var mat = Matrix.one(2, 3);

  print(mat);

  // Matrix: 2x3
  // [1.0, 1.0, 1.0]
  // [1.0, 1.0, 1.0]
}

/// Matrix of a certain number
void createNumbers() {
  var mat = Matrix.number(9, 2, 3);

  print(mat);

  // Matrix: 2x3
  // [9.0, 9.0, 9.0]
  // [9.0, 9.0, 9.0]
}

/// Matrix of random numbers
void createRandom() {
  var mat = Matrix.random(2, 3);

  print(mat);

  // Matrix: 2x3
  // [0.38224693703597046, 0.5412146597032305, 0.6424342222644003]
  // [0.8491145735932242, 0.6236773300386973, 0.25269555696856316]
}

/// Access item
void queryElement() {
  var mat = Matrix.fromList([
    [2, 3, 3, 3],
    [9, 9, 8, 6],
    [1, 1, 2, 9]
  ]);

  print(mat.matrix[1][2]);

  // 8.0
}

/// Access row
void queryRow() {
  var mat = Matrix.fromList([
    [2, 3, 3, 3],
    [9, 9, 8, 6],
    [1, 1, 2, 9]
  ]);

  print(mat.matrix[0]);

  // [2.0, 3.0, 3.0, 3.0]
}

/// Access column
void queryColumn() {
  var mat = Matrix.fromList([
    [2, 3, 3, 3],
    [9, 9, 8, 6],
    [1, 1, 2, 9]
  ]);

  print(mat.column(0));

  // [2.0, 9.0, 1.0]
}

/// Transpose
void transpose() {
  var mat = Matrix.fromList([
    [2, 3, 3, 3],
    [9, 9, 8, 6],
    [1, 1, 2, 9]
  ]);

  print(mat.transpose);

  // Matrix: 4x3
  // [2.0, 9.0, 1.0]
  // [3.0, 9.0, 1.0]
  // [3.0, 8.0, 2.0]
  // [3.0, 6.0, 9.0]
}

/// Row echelon form
void rowEchelonForm() {
  var mat = Matrix.fromList([
    [2, 3, 3, 3],
    [9, 9, 8, 6],
    [1, 1, 2, 9]
  ]);

  print(mat.rowEchelonForm);

  // Matrix: 3x4
  // [1.0, 0.0, 0.0, 1.5]
  // [0.0, 1.0, 0.0, -7.5]
  // [0.0, 0.0, 1.0, 7.5]
}

/// Rank
void rank() {
  var mat = Matrix.fromList([
    [2, 3, 3, 3],
    [9, 9, 8, 6],
    [1, 1, 2, 9]
  ]);

  print(mat.rank);

  // 3
}

/// Update item
void updateElement() {
  var mat = Matrix.fromList([
    [2, 3, 3, 3],
    [9, 9, 8, 6],
    [1, 1, 2, 9]
  ]);

  mat.matrix[0][0] = 3;

  print(mat.matrix[0]);

  // [3.0, 3.0, 3.0, 3.0]
}

/// Update row
void updateRow() {
  var mat = Matrix.fromList([
    [2, 3, 3, 3],
    [9, 9, 8, 6],
    [1, 1, 2, 9]
  ]);

  mat.matrix[0] = [1, 4, 5, 0];

  print(mat.matrix[0]);

  // [1.0, 4.0, 5.0, 0.0]
}

/// Update column
void updateColumn() {
  var mat = Matrix.fromList([
    [2, 3, 3, 3],
    [9, 9, 8, 6],
    [1, 1, 2, 9]
  ]);

  mat.setColumn([1, 4, 5], 0);

  print(mat);

  // Matrix: 3x4
  // [1.0, 3.0, 3.0, 3.0]
  // [4.0, 9.0, 8.0, 6.0]
  // [5.0, 1.0, 2.0, 9.0]
}

/// Insert row
void insertRow() {
  var mat = Matrix.fromList([
    [2, 3, 3, 3],
    [9, 9, 8, 6],
    [1, 1, 2, 9]
  ]);

  mat.addRow([8, 8, 8, 8], 0);

  print(mat);

  // Matrix: 4x4
  // [8.0, 8.0, 8.0, 8.0]
  // [2.0, 3.0, 3.0, 3.0]
  // [9.0, 9.0, 8.0, 6.0]
  // [1.0, 1.0, 2.0, 9.0]
}

void insertColumn() {
  var mat = Matrix.fromList([
    [2, 3, 3, 3],
    [9, 9, 8, 6],
    [1, 1, 2, 9]
  ]);

  mat.addColumn([8, 8, 8], 4);

  print(mat);

  // Matrix: 3x5
  // [2.0, 3.0, 3.0, 3.0, 8.0]
  // [9.0, 9.0, 8.0, 6.0, 8.0]
  // [1.0, 1.0, 2.0, 9.0, 8.0]
}

/// Append rows
void appendRows() {
  var mat = Matrix.fromList([
    [2, 3, 3, 3],
    [9, 9, 8, 6],
    [1, 1, 2, 9]
  ]);

  var tail = Matrix.fromList([
    [8, 8, 8, 8],
    [8, 8, 8, 8]
  ]);

  mat.appendRows(tail);

  print(mat);

  // Matrix: 5x4
  // [2.0, 3.0, 3.0, 3.0]
  // [9.0, 9.0, 8.0, 6.0]
  // [1.0, 1.0, 2.0, 9.0]
  // [8.0, 8.0, 8.0, 8.0]
  // [8.0, 8.0, 8.0, 8.0]
}

/// Append columns
void appendColumn() {
  var mat = Matrix.fromList([
    [2, 3, 3, 3],
    [9, 9, 8, 6],
    [1, 1, 2, 9]
  ]);

  var tail = Matrix.fromList([
    [8, 8],
    [8, 8],
    [8, 8]
  ]);

  mat.appendColumns(tail);

  print(mat);

  // Matrix: 3x6
  // [2.0, 3.0, 3.0, 3.0, 8.0, 8.0]
  // [9.0, 9.0, 8.0, 6.0, 8.0, 8.0]
  // [1.0, 1.0, 2.0, 9.0, 8.0, 8.0]
}

/// Delete row
void deleteRow() {
  var mat = Matrix.fromList([
    [2, 3, 3, 3],
    [9, 9, 8, 6],
    [1, 1, 2, 9]
  ]);

  mat.deleteRow(0);

  print(mat);

  // Matrix: 2x4
  // [9.0, 9.0, 8.0, 6.0]
  // [1.0, 1.0, 2.0, 9.0]
}

/// Delete column
void deleteColumn() {
  var mat = Matrix.fromList([
    [2, 3, 3, 3],
    [9, 9, 8, 6],
    [1, 1, 2, 9]
  ]);

  mat.deleteColumn(0);

  print(mat);

  // Matrix: 3x3
  // [3.0, 3.0, 3.0]
  // [9.0, 8.0, 6.0]
  // [1.0, 2.0, 9.0]
}

/// Delete rows
///
/// While deleting rows, the indexs MUST be put in order FROM SMALLEST TO LARGEST.
void deleteRows() {
  var mat = Matrix.fromList([
    [2, 3, 3, 3],
    [9, 9, 8, 6],
    [1, 1, 2, 9]
  ]);

  mat.deleteRows([0, 1]);

  print(mat);

  // Matrix: 1x4
  // [1.0, 1.0, 2.0, 9.0]
}

/// Delete columns
///
/// While deleting columns, the indexs MUST be put in order FROM SMALLEST TO LARGEST.
void deleteColumns() {
  var mat = Matrix.fromList([
    [2, 3, 3, 3],
    [9, 9, 8, 6],
    [1, 1, 2, 9]
  ]);

  mat.deleteColumns([0, 2]);

  print(mat);

  // Matrix: 3x2
  // [3.0, 3.0]
  // [9.0, 6.0]
  // [1.0, 9.0]
}

/// Addition
void addition() {
  var mat = Matrix.fromList([
    [2, 3, 3],
    [1, 2, 9]
  ]);

  print(mat + 3);

  print(mat + mat);
}

/// Subtraction
void subtraction() {
  var mat = Matrix.fromList([
    [2, 3, 3],
    [1, 2, 9]
  ]);

  print(mat - 3);
}

/// Scalar matrix multiplication
///
/// The scalar number MUST be in the right position
void scalarMultiplication() {
  var mat = Matrix.fromList([
    [2, 3, 3],
    [1, 2, 9]
  ]);

  print(mat * 3);

  // Matrix: 2x3
  // [6.0, 9.0, 9.0]
  // [3.0, 6.0, 27.0]
}

/// Matrix multiplication
void matrixMultiplication() {
  var mat = Matrix.fromList([
    [2, 3, 3],
    [1, 2, 9]
  ]);

  var multi = Matrix.fromList([
    [2, 3.5],
    [1, -2],
    [-4, 0.5]
  ]);

  print(mat * multi);

  // Matrix: 2x2
  // [-5.0, 2.5]
  // [-32.0, 4.0]
}

/// Create a square matrix
void createSquareMatrix() {
  var mat = SquareMatrix.fromList([
    [2, 3, 3],
    [1, 2, 9],
    [9, 8, 6]
  ]);

  print(mat);

  // Matrix: 3x3
  // [6.0, 9.0, 9.0]
  // [3.0, 6.0, 27.0]
  // [27.0, 24.0, 18.0]
}

/// SquareMatrix of zeros
void createSquareZeros() {
  var mat = SquareMatrix.zero(3);

  print(mat);

  // SquareMatrix: 3x3
  // [0.0, 0.0, 0.0]
  // [0.0, 0.0, 0.0]
  // [0.0, 0.0, 0.0]
}

/// SquareMatrix of ones
void createSquareOnes() {
  var mat = SquareMatrix.one(3);

  print(mat);

  // SquareMatrix: 3x3
  // [1.0, 1.0, 1.0]
  // [1.0, 1.0, 1.0]
  // [1.0, 1.0, 1.0]
}

/// SquareMatrix of a certain number
void createSquareNumbers() {
  var mat = SquareMatrix.number(2, 3);

  print(mat);

  // SquareMatrix: 3x3
  // [2.0, 2.0, 2.0]
  // [2.0, 2.0, 2.0]
  // [2.0, 2.0, 2.0]
}

/// SquareMatrix of random numbers
void createSquareRandoms() {
  var mat = SquareMatrix.random(3);

  print(mat);

  // SquareMatrix: 3x3
  // [0.2666268534804871, 0.607635195280705, 0.581392153915932]
  // [0.3590813328566256, 0.8098420655296595, 0.015949886001227154]
  // [0.6670401483574298, 0.3054732396044414, 0.3113168618967712]
}

/// Create diagonal square matrix from list
void diagFromList() {
  var mat = SquareMatrix.diagonal([1, 2, 3]);

  print(mat);

  // SquareMatrix: 3x3
  // [1.0, 0.0, 0.0]
  // [0.0, 2.0, 0.0]
  // [0.0, 0.0, 3.0]
}

/// Create diagonal square matrix from number
void diagFromNumber() {
  var mat = SquareMatrix.diagonalFromNumber(2, 3);

  print(mat);

  // SquareMatrix: 3x3
  // [2.0, 0.0, 0.0]
  // [0.0, 2.0, 0.0]
  // [0.0, 0.0, 2.0]
}

/// Determinant
void det() {
  var dett = SquareMatrix.fromList([
    [1, 2],
    [1, 1]
  ]);

  print(dett.determinant);

  // -1.0
}

/// Inverse
void inv() {
  var dett = SquareMatrix.fromList([
    [1, 2],
    [1, 1]
  ]);

  print(dett.inverse);

  // SquareMatrix: 2x2
  // [1.0, -1.0]
  // [1.0, -1.0]
}

/// Eigenvalues
void eig() {
  print("Unimplemented function.");
}

/// Eigenvectors
void eigVector() {
  print("Unimplemented function.");
}
